### ISSUE TITLE
add ReverseDiff

### ISSUE THREAD

**Issue by andreaskoher** [tpapp/LogDensityProblems.jl#75]
Hi @tpapp,

I really enjoy your DSL free approach in DynamicHMC. With Turing.jl I experienced substantial performance gains for medium sized problems using ReverseDiff and therefore, I would like to propose the following code snipped:

```julia
import ReverseDiff

struct ReverseDiffLogDensity{L, C} <: LogDensityProblems.ADGradientWrapper
    ℓ::L
    ctp::C
end

"""
    ADgradient(:ReverseDiff, ℓ)
    ADgradient(Val(:ReverseDiff), ℓ)

Gradient using algorithmic/automatic differentiation via ReverseDiff.
"""
LogDensityProblems.ADgradient(::Val{:ReverseDiff}, ℓ) = begin
	f = LogDensityProblems._logdensity_closure(ℓ)
	x = rand(dimension(ℓ.transformation))
	tape = ReverseDiff.GradientTape(f, x)
	compiledtape = ReverseDiff.compile(tape)
	ReverseDiffLogDensity(ℓ, compiledtape)
end

Base.show(io::IO, ∇ℓ::ReverseDiffLogDensity) = print(io, "ReverseDiff AD wrapper for ", ∇ℓ.ℓ)

function LogDensityProblems.logdensity_and_gradient(∇ℓ::ReverseDiffLogDensity, x::AbstractVector{T}) where {T}
    @unpack ℓ, compiledtape = ∇ℓ
    buffer = LogDensityProblems._diffresults_buffer(ℓ, x)
    result = ReverseDiff.gradient!(buffer, compiledtape, x)
    LogDensityProblems._diffresults_extract(result)
end
```

**Comment by andreaskoher** [comment]
and here is a simple benchmark example:

```julia
struct RandomWalkProblem{Ty}
    y::Ty
end

p = let 
    σ = 0.2
    x = 0:0.001:1
    y = sin.(2π*x) + rand(Normal(0,σ), length(x))
    RandomWalkProblem(y)
end

function (problem::RandomWalkProblem)(θ)
    @unpack σ, ŷ = θ               # extract the parameters
    @unpack y = problem

    ℓ  = 0.
    ℓ += logpdf(Uniform(-1,1), ŷ[1]) # prior for initial step  
    ℓ += logpdf(Exponential(0.2), σ) # noise prior
    ℓ += logpdf(Normal(ŷ[1], σ), y[1]) # likelihood

    for i in 2:length(y)
        ℓ += logpdf(Normal(ŷ[i-1], σ), ŷ[i]) # prior 
        ℓ += logpdf(Normal(ŷ[i], σ), y[i]) # likelihood
    end
    return ℓ
end

function problem_transformation(p::RandomWalkProblem)
    as((σ = asℝ₊, ŷ = as(Array, length(p.y))))
end

t = problem_transformation(p)
x = rand(dimension(t))
P = TransformedLogDensity(t, p)

using BenchmarkTools
∇P = ADgradient(:ForwardDiff, P)
@btime LogDensityProblems.logdensity_and_gradient(∇P, x) # 10ms

∇P = ADgradient(:Zygote, P)
@btime LogDensityProblems.logdensity_and_gradient(∇P, x) # 110s

∇P = ADgradient(Val(:ReverseDiff), P)
@btime LogDensityProblems.logdensity_and_gradient(∇P, x) # 1ms
```

**Comment by tpapp** [comment]
This is a good idea, a PR would be welcome.

**Comment by tpapp** [comment]
Fixed by #76, thanks!
